llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -0,0 +1,278 @@
+//===- AIEPostPipeliner.cpp - Post RA Pipeliner                            ===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+// This file contains a simple post-RA pipeliner. It tries to wrap the linear
+// schedule into a number of stages
+//===----------------------------------------------------------------------===//
+
+#include "AIEPostPipeliner.h"
+#include "Utils/AIELoopUtils.h"
+#include "llvm/Support/MathExtras.h"
+
+#define DEBUG_TYPE "postpipeliner"
+
+namespace llvm::AIE {
+
+PipelineScheduleVisitor::~PipelineScheduleVisitor() {}
+
+class PostPipelineDumper : public PipelineScheduleVisitor {
+public:
+  PostPipelineDumper() : PipelineScheduleVisitor() {}
+  void startPrologue() override { dbgs() << "Prologue:\n"; }
+  void startLoop() override { dbgs() << "Loop:\n"; }
+  void startEpilogue() override { dbgs() << "Epilogue:\n"; }
+  void startCycle() override { dbgs() << "\tBUNDLE {\n"; }
+  void emit(MachineInstr *MI) override { dbgs() << "\t\t" << *MI; }
+  void endCycle() override { dbgs() << "\t}\n"; }
+};
+
+PostPipeliner::PostPipeliner(const AIEHazardRecognizer &HR, int NInstr)
+    : HR(HR), NInstr(NInstr) {}
+
+bool PostPipeliner::accept(MachineBasicBlock &LoopBlock) {
+  // We leave the single-block loop criterion to our caller. It is fulfilled
+  // by being a loopaware scheduling candidate.
+  // First get us some instruments
+  const auto &ST = LoopBlock.getParent()->getSubtarget();
+  TII = static_cast<const AIEBaseInstrInfo *>(ST.getInstrInfo());
+
+  // 1. We need ZOL
+  auto Terminator = LoopBlock.getFirstInstrTerminator();
+  if (Terminator == LoopBlock.end() ||
+      !TII->isHardwareLoopEnd((*Terminator).getOpcode())) {
+    return false;
+  }
+  // 2. We need to fix the tripcount and to push out a time-critical prologue.
+  // If we don't have a dedicated preheader that is fallthrough, don't even
+  // bother.
+  Preheader = AIELoopUtils::getDedicatedFallThroughPreheader(LoopBlock);
+  if (!Preheader) {
+    return false;
+  }
+
+  // 3. We need to know how to update the tripcount
+  for (auto &MI : reverse(*Preheader)) {
+    if (TII->isZOLTripCountDef(MI)) {
+      TripCountDef = &MI;
+      break;
+    }
+  }
+  if (!TripCountDef) {
+    return false;
+  }
+
+  // 3a. If the addend immediate is non zero, we have been pipelined before
+  // and we can't trust min itercount metadata. Return on investment is probably
+  // low anyway.
+  // FIXME: Strictly speaking, this is AIE2 specific and should get some
+  // TII support. On the other hand, it will probably remain compatible.
+  if (TripCountDef->getNumOperands() != 3 ||
```

Comment:
Maybe something like:
```
TII->isNonPipelinedZOLTripCountDef(MI)
```
We can vanish some code here...

Reply:
Done. Included 'Pristine' parameter in isZOLTripCountDef